### PR TITLE
satellite/overlay: combine SelectStorageNodes and SelectNewStorageNodes

### PR DESCRIPTION
**Pull request by navillasa**
What:
- combines SelectStorageNodes and SelectNewStorageNodes

Why:
- consolidating the two functions to one means only one trip to the db

Please describe the tests:
 - TestEnsureMinimumRequested makes sure that the correct number of total nodes requested is returned, and that the correct number of reputable and new nodes are returned.
 - There are existing tests for node selection at satellite/overlay/service_test.go and satellite/overlay/selection_test.go. They ensure that the expected amounts of new nodes and/or reputable nodes are returned from the FindStorageNodes call.

Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [x] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

**Comment by egonelbre** [comment]
What's the performance impact of this?

**Comment by navillasa** [comment]
> What's the performance impact of this?

@egonelbre I'm not sure yet, looking into that today.

File: satellite/satellitedb/overlaycache.go [hunk 1]
```diff
@@ -33,112 +33,166 @@ type overlaycache struct {
...
-		for _, n := range moreNodes {
-			nodes = append(nodes, n)
-			criteria.ExcludedIDs = append(criteria.ExcludedIDs, n.Id)
-			criteria.ExcludedNetworks = append(criteria.ExcludedNetworks, n.LastNet)
-		}
-		if len(nodes) == count {
-			break
+	safeExcludeNodes := ""
+	if len(criteria.ExcludedIDs) > 0 {
+		safeExcludeNodes = ` AND id NOT IN (?` + strings.Repeat(", ?", len(criteria.ExcludedIDs)-1) + `)`
```
Comment (egonelbre) [inline satellite/satellitedb/overlaycache.go]: Let's use arrays rather than repeated `?`.

Comment (thepaul) [reply satellite/satellitedb/overlaycache.go]: I agree, but I think that should be a separate change. This is making a pretty significant change to pretty critical code, so I'd like to minimize the size of the diff.

File: satellite/satellitedb/overlaycache.go [hunk 2]
```diff
@@ -33,112 +33,166 @@ type overlaycache struct {
...
 	if !criteria.DistinctIP {
-		nodes, err = cache.queryNodes(ctx, criteria.ExcludedIDs, count, safeQuery, args...)
+		newNodeQuery, moreNewNodeArgs := buildSelection(ctx, newNodeCount, safeNewNodeQuery)
+		newNodeArgs = append(newNodeArgs, moreNewNodeArgs...)
+
+		reputableNodeQuery, moreReputableNodeArgs := buildSelection(ctx, reputableNodeCount, safeReputableNodeQuery)
+		reputableNodeArgs = append(reputableNodeArgs, moreReputableNodeArgs...)
+
+		finalQuery := newNodeQuery + " UNION ALL " + reputableNodeQuery
+		finalArgs := append(newNodeArgs, reputableNodeArgs...)
```
Comment (VinozzZ) [inline satellite/satellitedb/overlaycache.go]: Maybe we can initialize the `finalArgs` and then call append to that array? If someone adds another append after this line to newNodeArgs, we may have some problems

File: satellite/satellitedb/overlaycache.go [hunk 3]
```diff
@@ -33,112 +33,166 @@ type overlaycache struct {
...
+			newNodeQuery, moreNewNodeArgs = buildSelectionDistinct(ctx, criteria.ExcludedNetworks, newNodeCount, safeNewNodeQuery, true)
+			newNodeArgs = append(newNodeArgs, moreNewNodeArgs...)
+			// TODO: change how UNION ALL is added, depending on if we expect requests for only new nodes and not reputable nodes
+			newNodeQuery += "  UNION ALL "
+		}
+		reputableNodeQuery, moreReputableNodeArgs := buildSelectionDistinct(ctx, criteria.ExcludedNetworks, reputableNodeCount, safeReputableNodeQuery, false)
+		reputableNodeArgs = append(reputableNodeArgs, moreReputableNodeArgs...)
+
+		finalQuery := newNodeQuery + reputableNodeQuery
+		finalArgs := append(newNodeArgs, reputableNodeArgs...)
```
Comment (VinozzZ) [inline satellite/satellitedb/overlaycache.go]: same as above

File: satellite/overlay/benchmark_test.go [hunk 4]
```diff
@@ -302,31 +302,31 @@ func BenchmarkNodeSelection(b *testing.B) {
 
 		b.Run("SelectStorageNodes", func(b *testing.B) {
 			for i := 0; i < b.N; i++ {
-				selected, err := overlaydb.SelectStorageNodes(ctx, SelectCount, criteria)
+				selected, err := overlaydb.SelectStorageNodes(ctx, SelectCount, SelectCount, criteria)
```
Comment (egonelbre) [inline satellite/overlay/benchmark_test.go]: Shouldn't this be `SelectCount, 0`?

File: satellite/overlay/benchmark_test.go [hunk 5]
```diff
@@ -302,31 +302,31 @@ func BenchmarkNodeSelection(b *testing.B) {
...
+				selected, err := overlaydb.SelectStorageNodes(ctx, 0, SelectCount, criteria)
 				require.NoError(b, err)
 				require.NotEmpty(b, selected)
 			}
 		})
 
 		b.Run("SelectStorageNodesExclusion", func(b *testing.B) {
 			for i := 0; i < b.N; i++ {
-				selected, err := overlaydb.SelectStorageNodes(ctx, SelectCount, excludedCriteria)
+				selected, err := overlaydb.SelectStorageNodes(ctx, SelectCount, SelectCount, criteria)
```
Comment (egonelbre) [inline satellite/overlay/benchmark_test.go]: Shouldn't this be `SelectCount, 0`?

File: satellite/overlay/benchmark_test.go (hunk 4 above)
Comment (navillasa) [reply satellite/overlay/benchmark_test.go]: okay I can change this and the one below- I thought you wanted a case where both reputable storage nodes and new storage nodes were selected. not only cases where they're separate.

Comment (egonelbre) [reply satellite/overlay/benchmark_test.go]: These are specifically about measuring them separately. `FindStorageNodes` measures both together.

If you change what these benchmarks returns then they aren't comparable before and after, since they do different amount of work.

**Comment by navillasa** [comment]
```
name                                                              old time/op    new time/op    delta
NodeSelection/Postgres/Postgres/SelectStorageNodes-8                12.8ms ± 2%    14.1ms ±11%  +10.23%  (p=0.008 n=5+5)
NodeSelection/Postgres/Postgres/SelectNewStorageNodes-8             15.6ms ±13%    21.4ms ±35%     ~     (p=0.095 n=5+5)
NodeSelection/Postgres/Postgres/SelectStorageNodesExclusion-8       29.6ms ± 3%    33.6ms ±17%     ~     (p=0.151 n=5+5)
NodeSelection/Postgres/Postgres/SelectNewStorageNodesExclusion-8    32.4ms ±12%    44.3ms ±43%     ~     (p=0.222 n=5+5)
NodeSelection/Postgres/Postgres/FindStorageNodes-8                  72.9ms ±23%    54.1ms ±12%  -25.79%  (p=0.032 n=5+5)
NodeSelection/Postgres/Postgres/FindStorageNodesExclusion-8          188ms ±53%     164ms ±51%     ~     (p=0.690 n=5+5)

name                                                              old alloc/op   new alloc/op   delta
NodeSelection/Postgres/Postgres/SelectStorageNodes-8                 232kB ± 0%      38kB ± 0%  -83.75%  (p=0.008 n=5+5)
NodeSelection/Postgres/Postgres/SelectNewStorageNodes-8              229kB ± 0%      42kB ± 0%  -81.54%  (p=0.008 n=5+5)
NodeSelection/Postgres/Postgres/SelectStorageNodesExclusion-8        317kB ± 0%     127kB ± 0%  -59.83%  (p=0.008 n=5+5)
NodeSelection/Postgres/Postgres/SelectNewStorageNodesExclusion-8     316kB ± 0%     220kB ± 0%  -30.37%  (p=0.008 n=5+5)
NodeSelection/Postgres/Postgres/FindStorageNodes-8                   267kB ± 0%      65kB ± 0%  -75.52%  (p=0.008 n=5+5)
NodeSelection/Postgres/Postgres/FindStorageNodesExclusion-8          509kB ± 0%     342kB ± 0%  -32.80%  (p=0.008 n=5+5)

name                                                              old allocs/op  new allocs/op  delta
NodeSelection/Postgres/Postgres/SelectStorageNodes-8                 3.61k ± 0%     0.82k ± 0%  -77.25%  (p=0.008 n=5+5)
NodeSelection/Postgres/Postgres/SelectNewStorageNodes-8              3.21k ± 0%     0.85k ± 0%  -73.64%  (p=0.008 n=5+5)
NodeSelection/Postgres/Postgres/SelectStorageNodesExclusion-8        6.08k ± 0%     3.29k ± 0%  -45.82%  (p=0.008 n=5+5)
NodeSelection/Postgres/Postgres/SelectNewStorageNodesExclusion-8     5.68k ± 0%     5.88k ± 0%   +3.59%  (p=0.008 n=5+5)
NodeSelection/Postgres/Postgres/FindStorageNodes-8                   3.90k ± 0%     1.05k ± 0%  -73.18%  (p=0.008 n=5+5)
NodeSelection/Postgres/Postgres/FindStorageNodesExclusion-8          9.69k ± 0%     6.95k ± 0%  -28.30%  (p=0.008 n=5+5)
```

File: satellite/satellitedb/overlaycache.go [hunk 6]
```diff
@@ -33,112 +33,167 @@ type overlaycache struct {
...
+	WHERE disqualified IS NULL
+	AND suspended IS NULL
+	AND exit_initiated_at IS NULL
+	AND type = ?
+	AND free_disk >= ?
+	` + reputationCondition + `
+	AND last_contact_success > ?`
+	args = append(make([]interface{}, 0),
+		nodeType, criteria.FreeDisk, criteria.AuditCount,
+		criteria.UptimeCount, time.Now().Add(-criteria.OnlineWindow))
```
Comment (thepaul) [inline satellite/satellitedb/overlaycache.go]: It looks like, in the old code, criteria.UptimeCount always ended up being 0 when querying new nodes (it's not set in the NodeCriteria initialization at https://github.com/storj/storj/pull/3831/files#diff-5eccef64ca6f24b733402e9376a8f2fbL313 ). Is it correct to use a nonzero value here when `isNewNodeQuery` is true?

**Comment by egonelbre** [comment]
Note, your measurement error is quite high. Did you have other things running in the background?

Repeated the measurement locally:
```
go test -bench NodeSelection -run XXX -count 10 -benchtime 5s

>benchstat baseline.log combined.log
name                                                               old time/op  new time/op  delta
NodeSelection/Postgres/Postgres/SelectStorageNodes-32              6.67ms ± 1%  6.70ms ± 1%    ~     (p=0.247 n=10+10)
NodeSelection/Postgres/Postgres/SelectNewStorageNodes-32           7.05ms ± 0%  7.24ms ± 0%  +2.77%  (p=0.000 n=10+9)
NodeSelection/Postgres/Postgres/SelectStorageNodesExclusion-32     16.5ms ± 1%  16.0ms ± 0%  -3.12%  (p=0.000 n=8+9)
NodeSelection/Postgres/Postgres/SelectNewStorageNodesExclusion-32  16.5ms ± 1%  17.1ms ± 0%  +3.52%  (p=0.000 n=10+8)
NodeSelection/Postgres/Postgres/FindStorageNodes-32                45.3ms ± 2%  43.5ms ± 0%  -3.91%  (p=0.000 n=10+10)
NodeSelection/Postgres/Postgres/FindStorageNodesExclusion-32       76.2ms ± 0%  76.0ms ± 0%    ~     (p=0.436 n=10+10)
```

File: satellite/satellitedb/overlaycache.go [hunk 7]
```diff
@@ -33,112 +33,167 @@ type overlaycache struct {
...
+		tempNewNodeArgs := []interface{}{}
+		tempNewNodeArgs = append(tempNewNodeArgs, newNodeArgs...)
+		tempReputableNodeArgs := []interface{}{}
+		tempReputableNodeArgs = append(tempReputableNodeArgs, reputableNodeArgs...)
+
+		if receivedNewNodeCount < newNodeCount {
+			newNodeQuery, moreNewNodeArgs := buildSelectionDistinct(ctx, criteria.ExcludedNetworks, newNodeCount, safeNewNodeQuery, true)
+			tempNewNodeArgs = append(tempNewNodeArgs, moreNewNodeArgs...)
+			finalArgs = append(finalArgs, tempNewNodeArgs...)
+			finalQuery = newNodeQuery + "  UNION ALL "
```
Comment (egonelbre) [inline satellite/satellitedb/overlaycache.go]: Why does it need to do a union here?

File: satellite/satellitedb/overlaycache.go (hunk 6 above)
Comment (navillasa) [reply satellite/satellitedb/overlaycache.go]: good catch, I'll change that to be consistent with how it was before

File: satellite/satellitedb/overlaycache.go (hunk 7 above)
Comment (navillasa) [reply satellite/satellitedb/overlaycache.go]: the union all is so that the new node query string can be concatenated to the reputable node query string below. if you would prefer, I could move the union all to line 109.

Comment (egonelbre) [reply satellite/satellitedb/overlaycache.go]: At this part I don't think it should need to do a union any more since the initial node selection failed so it should fallback to a wider implementation.

Comment (navillasa) [reply satellite/satellitedb/overlaycache.go]: I'm not sure what you mean by "since the initial node selection failed" or "it should fallback to a wider implementation" - could you clarify?

Comment (egonelbre) [reply satellite/satellitedb/overlaycache.go]: The initial selection tries to get the necessary ratio of new vs old. However when that fails it should try to select either to fill the necessary limit.

Comment (egonelbre) [reply satellite/satellitedb/overlaycache.go]: I just realized that it's also in the old implementation and I cannot quite understand why.

Comment (egonelbre) [reply satellite/satellitedb/overlaycache.go]: Found the reason https://github.com/storj/storj/pull/1890/files#r284444400

File: satellite/satellitedb/overlaycache.go [hunk 8]
```diff
@@ -33,112 +33,170 @@ type overlaycache struct {
...
-		args = append(args, v.Major, v.Major, v.Minor, v.Minor, v.Patch)
+	}
+	safeReputableNodeQuery, reputableNodeArgs, err := buildConditions(ctx, criteria, false)
+	if err != nil {
+		return nil, err
 	}
 
 	if !criteria.DistinctIP {
-		nodes, err = cache.queryNodes(ctx, criteria.ExcludedIDs, count, safeQuery, args...)
+		newNodeQuery, moreNewNodeArgs := buildSelection(ctx, newNodeCount, safeNewNodeQuery)
```
Comment (egonelbre) [inline satellite/satellitedb/overlaycache.go]: This is mixing query building and query execution layers making this hard to follow.

File: satellite/satellitedb/overlaycache.go [hunk 9]
```diff
@@ -33,112 +33,170 @@ type overlaycache struct {
...
+	receivedNewNodeCount := 0
+	receivedNodeNetworks := make(map[string]struct{})
+
 	for i := 0; i < 3; i++ {
-		moreNodes, err := cache.queryNodesDistinct(ctx, criteria.ExcludedIDs, criteria.ExcludedNetworks, count-len(nodes), safeQuery, criteria.DistinctIP, args...)
+		finalQuery := ""
+		finalArgs := []interface{}{}
+
+		// temporary arg slices are needed for each iteration, otherwise we'll end up with too many args appended
+		tempNewNodeArgs := []interface{}{}
```
Comment (egonelbre) [inline satellite/satellitedb/overlaycache.go]: This is mixing query building and query execution layers making this hard to follow.

File: satellite/satellitedb/overlaycache.go [hunk 10]
```diff
@@ -33,112 +33,170 @@ type overlaycache struct {
...
 	nodeType := int(pb.NodeType_STORAGE)
 
-	safeQuery := `
-		WHERE disqualified IS NULL
-		AND suspended IS NULL
-		AND exit_initiated_at IS NULL
-		AND type = ?
-		AND free_disk >= ?
+	// initiated with reputable node conditions
+	reputationCondition := `
```
Comment (egonelbre) [inline satellite/satellitedb/overlaycache.go]: The condition query and the arguments are spread apart making this hard to follow.

File: satellite/satellitedb/overlaycache.go [hunk 11]
```diff
@@ -33,112 +33,170 @@ type overlaycache struct {
...
+
+	safeNodeQuery := `
+	WHERE disqualified IS NULL
+	AND suspended IS NULL
+	AND exit_initiated_at IS NULL
+	AND type = ?
+	AND free_disk >= ?
+	` + reputationCondition + `
+	AND last_contact_success > ?`
+	args = append(make([]interface{}, 0),
```
Comment (egonelbre) [inline satellite/satellitedb/overlaycache.go]: why not just `[]interface{}{nodeType, criteria.FreeDisk ... }`

File: satellite/satellitedb/overlaycache.go [hunk 12]
```diff
@@ -167,65 +225,26 @@ func (cache *overlaycache) GetNodesNetwork(ctx context.Context, nodeIDs []storj.
...
-	}
-
-	safeExcludeNodes := ""
-	if len(excludedNodes) > 0 {
-		safeExcludeNodes = ` AND id NOT IN (?` + strings.Repeat(", ?", len(excludedNodes)-1) + `)`
-		for _, id := range excludedNodes {
-			args = append(args, id.Bytes())
-		}
-	}
+func buildSelection(ctx context.Context, count int, safeQuery string) (query string, args []interface{}) {
```
Comment (egonelbre) [inline satellite/satellitedb/overlaycache.go]: The query is returned even when count is 0.

File: satellite/satellitedb/overlaycache.go [hunk 13]
```diff
@@ -33,112 +33,170 @@ type overlaycache struct {
...
+	uptimeCount := criteria.UptimeCount
+
+	if isNewNodeQuery {
+		reputationCondition = `
 		AND (total_audit_count < ? OR total_uptime_count < ?)
-		AND last_contact_success > ?`
-	args := append(make([]interface{}, 0, 10),
-		nodeType, criteria.FreeDisk, criteria.AuditCount, criteria.UptimeCount, time.Now().Add(-criteria.OnlineWindow))
+		`
+		uptimeCount = 0
```
Comment (egonelbre) [inline satellite/satellitedb/overlaycache.go]: Is there a case where `total_uptime_count < 0`?

**Comment by JessicaGreben** [comment]
@egonelbre curious your thoughts... these changes seem to work and improve speed and doesn't vastly change the actual sql from implementation before. Can we merge this for the performance enhancments, then make another PR to fix your suggestions?

File: satellite/satellitedb/overlaycache.go (hunk 12 above)
Comment (thepaul) [reply satellite/satellitedb/overlaycache.go]: That should be fine: it ends up with `LIMIT 0` which lets the planner know that this part of the query can be ignored, and avoids the need to have a logic in the calling function to decide how to combine the two parts of the query (`UNION ALL` or not). And the execution time to build the query here is most likely dwarfed by the time it takes to execute the query. But it could be done either way, if you feel strongly about it.

Comment (egonelbre) [reply satellite/satellitedb/overlaycache.go:228]: The benchmark does not support that it's completely ignored. Maybe there's something else with the query, but there's a time increase for SelectStorageNodes and SelectNewStorageNodes in the parallel case.

**Comment by egonelbre** [comment]
@JessicaGreben I do not think the code is of sufficient quality here anymore. It wasn't great before, but now it's worse. I don't think 3% improvement is going to make a significant difference for the server.

File: satellite/satellitedb/overlaycache.go (hunk 12 above)
Comment (thepaul) [reply satellite/satellitedb/overlaycache.go:228]: You're right. I went back to verify. The planner doesn't ignore a subquery with `LIMIT 0`. I'm surprised, but there it is. Looks like we should indeed elide these when count=0.

File: satellite/satellitedb/overlaycache.go [hunk 14]
```diff
@@ -33,241 +33,269 @@ type overlaycache struct {
...
+			excludedNetworks := " AND last_net NOT IN (?" + strings.Repeat(", ?", len(criteria.ExcludedIDs)-1) + ") "
+			cond.addQuery(excludedNetworks)
+			for _, subnet := range criteria.ExcludedNetworks {
+				cond.addArg(subnet)
+			}
 		}
+		cond.addQuery(
+			`AND last_net <> ''
+			ORDER BY last_net, RANDOM()
+			) filteredcandidates`)
```
Comment (egonelbre) [inline satellite/satellitedb/overlaycache.go:192]: `ORDER BY last_net, RANDOM() ) filteredcandidates`, this isn't really a condition

**Comment by navillasa** [comment]
benchstat old_bench.txt new_bench.txt
```
name                                                              old time/op  new time/op  delta
NodeSelection/Postgres/Postgres/SelectStorageNodes-8              5.50ms ± 0%  5.49ms ± 0%    ~     (p=0.278 n=10+9)
NodeSelection/Postgres/Postgres/SelectNewStorageNodes-8           5.69ms ± 0%  5.35ms ± 0%  -6.00%  (p=0.000 n=9+9)
NodeSelection/Postgres/Postgres/SelectStorageNodesExclusion-8     14.8ms ± 0%  14.9ms ± 0%  +0.71%  (p=0.000 n=10+10)
NodeSelection/Postgres/Postgres/SelectNewStorageNodesExclusion-8  15.0ms ± 0%  14.8ms ± 0%  -1.35%  (p=0.000 n=10+10)
NodeSelection/Postgres/Postgres/FindStorageNodes-8                96.5ms ± 0%  92.3ms ± 0%  -4.30%  (p=0.000 n=9+8)
NodeSelection/Postgres/Postgres/FindStorageNodesExclusion-8        127ms ± 0%   123ms ± 1%  -3.18%  (p=0.000 n=9+10)
```

**Comment by stefanbenten** [comment]
Not much of an improvement, besides half the queries, right?

**Comment by thepaul** [comment]
> Not much of an improvement, besides half the queries, right?

Yep, the only benefit to expect from this is fewer queries. I would not expect that to be a very noticeable difference when benchmarked against a dedicated db on localhost, though, where RTT is very small. In our prod environments I'd expect a significantly bigger win. Big enough to justify the change? That I don't know. But in general, reducing db RTTs is worthwhile.

File: satellite/satellitedb/nodeselection.go [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+	}
+	if criteria.DistinctIP {
+		if len(excludedNetworks) > 0 {
+			excludedNetworksQuery := " AND last_net NOT IN (?" + strings.Repeat(", ?", len(excludedIDs)-1) + ") "
+			cond.addQuery(excludedNetworksQuery)
+			for _, subnet := range excludedNetworks {
+				cond.addArg(subnet)
+			}
+		}
+		cond.addQuery("AND last_net <> ''")
```
Comment (egonelbre) [inline satellite/satellitedb/nodeselection.go:191]: add whitespace in the beginning of this query

Comment (navillasa) [reply satellite/satellitedb/nodeselection.go]: ok added

File: satellite/satellitedb/nodeselection.go [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+	fmt.Fprintf(&q, " WHERE %s ", partial.condition.query)
+	args = append(args, partial.condition.args...)
+
+	if partial.orderBy != "" {
+		fmt.Fprintf(&q, " ORDER BY %s, RANDOM() ", partial.orderBy)
+	} else {
+		fmt.Fprint(&q, " ORDER BY RANDOM() ")
+	}
+
+	fmt.Fprintf(&q, " LIMIT ? ")
```
Comment (egonelbre) [inline satellite/satellitedb/nodeselection.go:215]: When it's doing a distinct query, it shouldn't limit here.

File: satellite/overlay/service.go [hunk 17]
```diff
@@ -342,14 +315,11 @@ func (service *Service) FindStorageNodesWithPreferences(ctx context.Context, req
 		OnlineWindow:     preferences.OnlineWindow,
 		DistinctIP:       preferences.DistinctIP,
 	}
-	reputableNodes, err := service.db.SelectStorageNodes(ctx, reputableNodeCount-len(newNodes), &criteria)
+	nodes, err = service.db.SelectStorageNodes(ctx, reputableNodeCount, newNodeCount, &criteria)
```
Comment (egonelbre) [inline satellite/overlay/service.go:318]: Let's add a test for SelectStorageNodes where `reputableNodeCount = N` and `newNodeCount = K` and there are `N + K` total nodes where all of them are reputable. In that case it should end up selecting all of the reputable nodes.

Comment (egonelbre) [reply satellite/overlay/service.go:318]: And a case where "reputable node count = N" and "new node count = K" and ensure that it only ends up selecting MinimumRequiredNodes or RequestedCount. It feels like it won't return the same number of nodes as it did before.

Comment (egonelbre) [reply satellite/overlay/service.go:318]: `reputableNodeCount` here is actually "required node count", how many nodes to return.

Comment (navillasa) [reply satellite/overlay/service.go:318]: For your first comment, this test exists already here, and the code passes this test: https://github.com/storj/storj/blob/master/satellite/overlay/selection_test.go#L198
I added TestEnsureMinimumRequested for the other concern.

Comment (egonelbre) [reply satellite/overlay/service.go:318]: The `TestEnsureMinimumRequested` had few bugs in it, I fixed them and now it is failing, whereas with the old code it passes.

**Comment by navillasa** [comment]
moved TestEnsureMinimumRequired to here to reduce size of the diff https://review.dev.storj.io/c/storj/storj/+/1468

File: satellite/satellitedb/nodeselection.go [hunk 18]
```diff
@@ -0,0 +1,323 @@
...
+
+func union(q1, q2 query) query {
+	if isEmpty(q1) {
+		return q2
+	}
+	if isEmpty(q2) {
+		return q1
+	}
+
+	finalQuery := "(" + q1.query + ") UNION ALL (" + q2.query + ")"
```
Comment (JessicaGreben) [inline satellite/satellitedb/nodeselection.go:265]: why `union all` versus just `union`?

Comment (navillasa) [reply satellite/satellitedb/nodeselection.go:265]: since `union` removes duplicates but `union all` does not, I thought it would be better to use `union all` for the very slight performance gain and because there shouldn't be any duplicates between them anyway. https://www.postgresqltutorial.com/postgresql-union/

File: satellite/satellitedb/nodeselection.go [hunk 19]
```diff
@@ -0,0 +1,323 @@
...
+	"storj.io/storj/satellite/overlay"
+)
+
+func (cache *overlaycache) SelectStorageNodes(ctx context.Context, totalNeededNodes, newNodeCount int, criteria *overlay.NodeCriteria) (nodes []*overlay.SelectedNode, err error) {
+	defer mon.Task()(&ctx)(&err)
+	if totalNeededNodes == 0 {
+		return nil, nil
+	}
+
+	needNewNodes := newNodeCount
```
Comment (mobyvb) [inline satellite/satellitedb/nodeselection.go:27]: should there be a check to make sure that `newNodeCount` is less than `totalNeededNodes`?

File: satellite/satellitedb/nodeselection.go [hunk 20]
```diff
@@ -0,0 +1,323 @@
...
+		}
+	}
+	return reputableNodes, newNodes, nil
+}
+
+func buildConditions(ctx context.Context, criteria *overlay.NodeCriteria, excludedIDs []storj.NodeID, excludedNetworks []string, isNewNodeQuery bool) (condition, error) {
+	var conds conditions
+	if isNewNodeQuery {
+		conds.add(
+			"total_audit_count < ?",
```
Comment (mobyvb) [inline satellite/satellitedb/nodeselection.go:155]: we should probably add `OR total_uptime_count < ?` so that it is consistent with the reputation criteria for vetted nodes
(@jenlij )

File: satellite/satellitedb/nodeselection.go [hunk 21]
```diff
@@ -0,0 +1,323 @@
...
+	if len(excludedIDs) > 0 {
+		excludedIDsQuery := " AND id NOT IN (?" + strings.Repeat(", ?", len(excludedIDs)-1) + ") "
+		cond.addQuery(excludedIDsQuery)
+		for _, id := range excludedIDs {
+			cond.addArg(id)
+		}
+	}
+	if criteria.DistinctIP {
+		if len(excludedNetworks) > 0 {
+			excludedNetworksQuery := " AND last_net NOT IN (?" + strings.Repeat(", ?", len(excludedIDs)-1) + ") "
```
Comment (mobyvb) [inline satellite/satellitedb/nodeselection.go:193]: `len(excludedNetworks)` instead of `len(excludedIDs)`?

File: satellite/satellitedb/nodeselection.go [hunk 22]
```diff
@@ -0,0 +1,323 @@
...
+		}
+		if lastIPPort.Valid {
+			node.LastIPPort = lastIPPort.String
+		}
+		if isNew {
+			newNodes = append(newNodes, &node)
+		} else {
+			reputableNodes = append(reputableNodes, &node)
+		}
+		if len(newNodes)+len(reputableNodes) == reputableNodeCount+newNodeCount {
```
Comment (mobyvb) [inline satellite/satellitedb/nodeselection.go:144]: I would prefer `if len(newNodes) >= newNodeCount && len(reputableNodes) >= reputableNodeCount`

File: satellite/satellitedb/nodeselection.go (hunk 19 above)
Comment (navillasa) [reply satellite/satellitedb/nodeselection.go:27]: that's a good idea!

File: satellite/satellitedb/nodeselection.go (hunk 20 above)
Comment (navillasa) [reply satellite/satellitedb/nodeselection.go:155]: so the reason I left this out is because in the original code at the satellite/overlay/service.go level, total uptime count isn't even passed in as a selection criteria: https://github.com/storj/storj/blob/master/satellite/overlay/service.go#L315

so right now, the `total_uptime_count` will always be zero, which means that only the audit count check will matter at this filter: https://github.com/storj/storj/blob/master/satellite/satellitedb/overlaycache.go#L102

File: satellite/satellitedb/nodeselection.go (hunk 21 above)
Comment (navillasa) [reply satellite/satellitedb/nodeselection.go:193]: oh shoot good catch 😬

File: satellite/satellitedb/nodeselection.go (hunk 22 above)
Comment (navillasa) [reply satellite/satellitedb/nodeselection.go:144]: okay cool I can change that

File: satellite/satellitedb/nodeselection.go (hunk 20 above)
Comment (mobyvb) [reply satellite/satellitedb/nodeselection.go:155]: Yeah, I am aware of the inconsistency in the original code - I was discussing it with @jenlij yesterday. But I think it would be better to include total_uptime_count, even if it is zero just so we have consistency.

Either way, Jen is working on a change to have a `vetted` flag, which will be used instead of comparing against total_audit_count and total_uptime_count anyway, so it's not a big deal either way. I don't consider this to be an approval-blocker.

File: satellite/satellitedb/nodeselection.go [hunk 23]
```diff
@@ -0,0 +1,327 @@
...
+		}
+	}
+	return nodes, nil
+}
+
+func (cache *overlaycache) selectStorageNodesOnce(ctx context.Context, reputableNodeCount, newNodeCount int, criteria *overlay.NodeCriteria, excludedIDs []storj.NodeID, excludedNetworks []string) (reputableNodes, newNodes []*overlay.SelectedNode, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	newNodeSelection := `SELECT last_net, id, address, last_ip_port, true FROM nodes`
+	reputableNodeSelection := `SELECT last_net, id, address, last_ip_port, false FROM nodes`
```
Comment (JessicaGreben) [inline satellite/satellitedb/nodeselection.go:94]: whats the false about here?

Comment (navillasa) [reply satellite/satellitedb/nodeselection.go:94]: when the query is run, it returns both new and reputable nodes together, and I wasn't sure of an easy way to distinguish the two. the bool here represents `isNewNode`- so when we scan the rows later, we can scan this value into a bool to determine what kind of node it is. it's weird. I can add a comment for clarification.

File: satellite/satellitedb/nodeselection.go [hunk 24]
```diff
@@ -0,0 +1,329 @@
...
+		if needNewNodes > 0 && receivedNewNodes == 0 {
+			needReputableNodes += needNewNodes
+			needNewNodes = 0
+		}
+
+		if needReputableNodes <= 0 && needNewNodes <= 0 {
+			break
+		}
+	}
+	return nodes, nil
```
Comment (mobyvb) [inline satellite/satellitedb/nodeselection.go:87]: nit - should there be a final check to test if len(nodes) is at least the amount requested, and return an error otherwise? Or is the behavior we want to return the list of nodes, even if there are not enough?

Comment (navillasa) [reply satellite/satellitedb/nodeselection.go:87]: hmm I see in the existing code that we return the final amount without checking length and do a `if len(nodes) < reputableNodeCount` check in the overlay/service level to return an error from FindStorageNodesWithPreferences. I'm kind of fine with leaving that as is, unless you think it would be better to return the error from this function.

Comment (mobyvb) [reply satellite/satellitedb/nodeselection.go:87]: oh yeah I see that it is happening in service.go. That is fine to me.
